File: backend/php/src/Services/AIAgentService.php
<?php
// SOURCE: CIE_Master_Developer_Build_Spec.docx §7 — AI Agent specificity check

namespace App\Services;

/**
 * Deterministic specificity check for expert authority statements.
 *
 * Verifies that a given text references a recognised standard, certification,
 * or technical specification (e.g. BS 7671, CE, UKCA, IP rating).
 * Rejects generic marketing phrases that carry no verifiable authority.
 */
class AIAgentService
{
    private const STANDARD_PATTERN = '/(?:'
        . '\bBS\s*\d+'
        . '|\bISO\s*\d+'
        . '|\bEN\s*\d+'
        . '|\bIEC\s*\d+'
        . '|\bIEEE\s*\d+'
        . '|\bANSI\b'
        . '|\bUL\s*\d+'
        . '|\bNEC\b'
        . '|\bATEX\b'
        . '|\bRoHS\b'
        . '|\bREACH\b'
        . '|\bUKCA\b'
        . '|\bCE\b'
        . '|\bIP\d{2}\b'
        . '|\bClass\s+[I12]\b'
        . '|\bRated\s+to\b'
        . '|\b\d+\s*[AW]\s*\/\s*\d+\s*[AW]\b'
        . ')/i';

    /**
     * Return true when $text references a recognised standard, certification,
     * or rated specification. Return false for empty or generic content.
     */
    public static function checkSpecificity(string $text): bool
    {
        $trimmed = trim($text);
        if ($trimmed === '') {
            return false;
        }

        return (bool) preg_match(self::STANDARD_PATTERN, $trimmed);
    }
}

File: backend/python/api/ai_agent_service.py
# SOURCE: CIE_Master_Developer_Build_Spec.docx §7 — AI Agent specificity check
"""
Deterministic specificity check for expert authority statements.

Verifies that expert_authority references a recognised standard,
certification, or technical specification (e.g. BS 7671, CE, UKCA, IP rating).
Rejects generic marketing phrases that carry no verifiable authority.
"""
from __future__ import annotations

import re

_STANDARD_PATTERN = re.compile(
    r'(?:'
    r'\bBS\s*\d+'
    r'|\bISO\s*\d+'
    r'|\bEN\s*\d+'
    r'|\bIEC\s*\d+'
    r'|\bIEEE\s*\d+'
    r'|\bANSI\b'
    r'|\bUL\s*\d+'
    r'|\bNEC\b'
    r'|\bATEX\b'
    r'|\bRoHS\b'
    r'|\bREACH\b'
    r'|\bUKCA\b'
    r'|\bCE\b'
    r'|\bIP\d{2}\b'
    r'|\bClass\s+[I12]\b'
    r'|\bRated\s+to\b'
    r'|\b\d+\s*[AW]\s*/\s*\d+\s*[AW]\b'
    r')',
    re.IGNORECASE,
)


def check_specificity(text: str) -> bool:
    """Return True when *text* references a recognised standard, certification,
    or rated specification. Return False for empty or generic content."""
    if not text or not text.strip():
        return False
    return bool(_STANDARD_PATTERN.search(text))
